src/ssl/bio.cc

```diff
@@ -324,17 +346,83 @@ Ssl::ServerBio::readAndParse(char *buf, const int size, BIO *table)
 /// Reads more data into the read buffer. Returns either the number of bytes
 /// read or, on errors (including "try again" errors), a negative number.
 int
-Ssl::ServerBio::readAndBuffer(BIO *table)
+Ssl::ServerBio::readAndBuffer(BIO *table, const int size)
 {
-    char *space = rbuf.rawAppendStart(SQUID_TCP_SO_RCVBUF);
-    const int result = Ssl::Bio::read(space, SQUID_TCP_SO_RCVBUF, table);
+    char *space = rbuf.rawAppendStart(size);
+    const int result = Ssl::Bio::read(space, size, table);
     if (result <= 0)
         return result;
 
     rbuf.rawAppendFinish(space, result);
     return result;
 }
 
+#if OPENSSL_KTLS_SUPPORT
+/// Read and give everything to our parser. (KTLS support ver.)
+/// When/if parsing is finished (successfully or not), start giving to OpenSSL.
+///
+/// \note If using KTLS, we must not consume more socket data than the caller (OpenSSL) expects.
+/// Otherwise, the session establishment potentially fails depending on the timing of packet arrival.
+/// Therefore, we peek the socket here and ensure we do not consume the socket data more than required.
```

Comment:
**This comment is explaining what I have understood so far and it might be wrong.**

> "all KTLS-supporting implementations must peek at the socket"

This part is perhaps not true.
General implementation just needs to use SSL_* with the regular BIO.
Squid might be some exception because it has to intercept the traffic using the wrapper BIO.
I agree that the code is a kind of workaround to satisfy squid's current requirement.

I was searching if there is any technical reference.
So far, what I could find is only this one that is telling reading the socket after negotiation might be undefined behaviour in FreeBSD.

https://gist.github.com/vladak/8fc4bb65f68a55eb98630b5ab5c6a4b9

> receive needs to be done using recvmsg(2)
> Once TLS receive is enabled by a successful set of the TCP_RXTLS_ENABLE socket option, all data read from the socket is returned as decrypted TLS records. Each received TLS record must be read from the socket using recvmsg(2). Each received TLS record will contain a TLS_GET_RECORD control message along with the decrypted payload. The control message contains a struct tls_get_record which includes fields from the TLS record header. If an invalid or corrupted TLS record is received, recvmsg(2) will fail with one of the following errors...
> 
> TODO: what happens if the app calls read(2) on the socket ?

Reply:
I found some related OpenSSL code and [discussion](https://github.com/openssl/openssl/pull/7848#discussion_r245650388). AFAICT, the situation is more nuanced (and arguably worse) than I expected:

* OpenSSL has a `read_ahead` mode that enables reading more socket bytes than OpenSSL needs to parse the next TLS element. If that mode is enabled, OpenSSL will usually request more socket bytes than the handshake. I speculate that `read_ahead` applies to handshake reading/parsing. Thus, our attempts to prevent "read past the handshake" by relying on OpenSSL to tell us how much to reach will fail.
* Since 2022, OpenSSL [enables](https://github.com/openssl/openssl/blob/0a22436ea5826d0089db7f1cd97b7c90135ca165/ssl/record/methods/ktls_meth.c#L439C14-L442) read_ahead mode when KTLS is in use: "With KTLS we always try and read as much as possible and fill the buffer". Thus, our attempts to prevent "read past the handshake" appear to work against OpenSSL desires.
* Kernel KTLS API (SOL_TLS, tls12_crypto_info_aes_gcm_128 and friends) probably does not provide the ability to report extra bytes read to the kernel (extra bytes read while parsing the handshake), especially bytes from an incomplete TLS record. The kernel may assume that it starts reading from the TLS record boundary.

Combining all of the above, I _speculate_ that OpenSSL will often attempt to read more bytes than the size of TLS handshake (while processing that handshake and before giving decryption control to the kernel) and, hence, will deal with any extra bytes read. When KTLS is enabled for a TLS connection, OpenSSL may assume that any "extra" bytes will not split a TLS record (because initial TLS records are usually not split across socket writes or reads?).

I failed to find OpenSSL documentation that outlines requirements for KTLS-compatible BIOs.

Based on the above, I hope we do not need to treat KTLS I/Os specially in our custom BIO, but, AFAICT, there is currently not enough information available to confirm or reject that conclusion.

Reply:
> reading the socket after negotiation might be undefined behaviour

Yes, that makes a lot of sense: There can be only one socket reader in this context. Once kernel starts reading the socket, the user must not.

What we do not know is whether it is OK for BIO, _during negotiation_, to read more bytes than OpenSSL has asked for. I suspect the answer depends on whether OpenSSL has code to get the buffered data from our BIO (kind of "flushing" our read BIO buffers -- there is OpenSSL code to flush write BIO buffers IIRC). I suspect that OpenSSL code does not exist and, hence, we cannot read more than OpenSSL asked (even if OpenSSL itself asks to read more than the handshake!) because those buffered bytes will be effectively lost, breaking TLS.

Reply:
If record_ is set before negotiation finishes (it look like so), readAndBuffer(table, SQUID_TCP_SO_RCVBUF)  must be readAndBuffer(table, size) (for KTLS).

I might've misread it before.

```
/// Read and give everything to OpenSSL.
int
Ssl::ServerBio::readAndGive(char *buf, const int size, BIO *table)
{
    // If we have unused buffered bytes, give those bytes to OpenSSL now,
    // before reading more. TODO: Read if we have buffered less than size?
    if (rbufConsumePos < rbuf.length())
        return giveBuffered(buf, size);

    if (record_) {
        const int result = readAndBuffer(table, SQUID_TCP_SO_RCVBUF);
        if (result <= 0)
            return result;
        return giveBuffered(buf, size);
    }

    return Ssl::Bio::read(buf, size, table);
}
```

Reply:
By this commit https://github.com/squid-cache/squid/pull/1629/commits/fa5d875cfb5444425bcdaecdedafa8a0e2e998e2,

> during negotiation, to read more bytes than OpenSSL has asked for.

"read more bytes than OpenSSL during cipher negotiation" is not expected to occur any more. (read more bytes than OpenSSL occurs but only for Hello.)


> (even if OpenSSL itself asks to read more than the handshake!)

This is difficult, but if it really happens, the native OpenSSL socket BIO should have the same issue.